incubator/hnc/pkg/reconcilers/hnc_config.go

```diff
@@ -162,6 +162,29 @@ func (r *ConfigReconciler) syncObjectReconcilers(ctx context.Context, inst *api.
 		}
 	}
 
+	// If a type exists in the forest but not exists in the Spec, we will
+	// set the mode of corresponding object reconciler to "ignore".
+	// TODO: Ideally, we should shut down the corresponding object
+	// reconciler. Gracefully terminating an object reconciler is still under
+	// development (https://github.com/kubernetes-sigs/controller-runtime/issues/764).
+	// We will revisit the code below once the feature is released.
+	for _, ts := range r.Forest.GetTypeSyncers() {
+		exist := false
+		for _, t := range inst.Spec.Types {
+			if ts.GetGVK() == schema.FromAPIVersionAndKind(t.APIVersion, t.Kind) {
+				exist = true
+				break
+			}
+		}
+		if exist {
+			continue
+		}
+		// The type does not exist in the Spec. Ignore subsequent reconciliations.
+		if err := ts.SetMode(ctx, api.Ignore, r.Log); err != nil {
+			return err // retry the reconciliation
+		}
+	}
+
```

Comment:
Just another thought from how HC did it but your call whether it's clearer either way.

Have a `map[TypeSyncer]bool` to represent if the `ts` exists in forest. Traverse `inst.Spec.Types` to do L154-163 and remove `ts` from map. Then traverse the `ts` remained in the map to do L183-185.

Reply:
Thanks for the suggestions! It will definitely help us get rid of an extra loop, but as a tradeoff, we need to store an additional map in the forest. I feel the extra loop might be ok in this case because we are not likely to have too many types to configure. But if it is not the case, or if we find it causes performance issues in future, I will refactor it as you suggested in a later PR.

Reply:
You could also build the map each time instead of storing it. Turns O(n^2) into O(n). But the types are probably small enough that big-O doesn't really matter for now